Week2/todo_app/program.js

```diff
@@ -16,43 +16,139 @@ class TodoModel {
         // Write code and query to delete an existing TODO item
+        const deleteQuery = `
+        DELETE FROM todo_items WHERE id = ${id}
+        `
+        ;
+        this.dbConnection.query(deleteQuery, (err,results, fields) => {
+            if(err) {
+                callback(err);
+                return;
+            } else {
+                callback(null, results);
+            }
+        })
+
     }
 
     tagTodoItem(todoItemId, tagId, callback) {
         // Write code and query add a tag to a TODO item
+        const tagTodoItemQuery = `
+        INSERT INTO todo_item_tag(todo_item_id, tag_id)
+        VALUES (${todoItemId},${tagId})
+        `
+        ;
+        this.dbConnection.query(tagTodoItemQuery, (err,results, fields) => {
+            if(err) {
+                callback(err);
+                return;
+            } else {
+                callback(null, results);
+            }
+        })
     }
         
     untagTodoItem(todoItemId, tagId, callback) {
         // Write code and query remove a tag from a TODO item
+        const untagTodoItemQuery = `
+        DELETE FROM todo_item_tag
+        WHERE todo_item_id = ${todoItemId}
+        `
+        ;
+        this.dbConnection.query(untagTodoItemQuery, (err,results, fields) => {
+            if(err) {
+                callback(err);
+                return;
+            } else {
+                callback(null, results);
+            }
+        })
     }
 
     markCompleted(todoItemId, callback) {
         // Write code to mark a TODO item as completed
+        const markCompletedQuery = `
+        UPDATE todo_items 
+        SET is_completed = 1 
+        WHERE id = ${todoItemId}
+        `
+        ;
+        this.dbConnection.query(markCompletedQuery, (err,results, fields) => {
+            if(err) {
+                callback(err);
+                return;
+            } else {
+                callback(null, results);
+            }
+        })
     }
+    help(callback) {
+        console.log(`
+            Hi! Please check the details and try again!
+
+            load                         ======>>>>> shows current todo list
+            create 'text' id             ======>>>>> creates a new todo item
+            update 'text' todo_id        ======>>>>> updates an existing todo item
+            delete todo_id               ======>>>>> deletes an existing todo item
+            tag    todo_item_id tag_id   ======>>>>> tags with an existing tag
+            untag  todo_item_id tag_id   ======>>>>> untags todo item
+            completed todo_item_id       ======>>>>> marks todo item item as 'done'
+            `
+        );
+    };
 }
 
 const dbConnection = mysql.createConnection({
     host     : 'localhost',
-    user     : 'root',
-    password : '',
+    user     : config.user,
+    password : config.password,
     database : 'todo_app'
 });
 
+const method = process.argv[2];
+const thirdInput = dbConnection.escape(process.argv[3]);
```

Comment:
Ah, I see - you do escaping here.
Not every value needs to be escaped, so it's probably better to do it nearer to where it's being used. If you find you have to repeat a lot of code, you could also use a helper function to make it more 'DRY'.